components/Donation/SingleProject/SingleProjectDonateModal.tsx

```diff
@@ -0,0 +1,370 @@
+"use client";
+
+import React, { useCallback, useMemo, useState } from "react";
+import toast from "react-hot-toast";
+import { type Hex, isAddress } from "viem";
+import { useAccount, useChainId, useSwitchChain } from "wagmi";
+import { Button } from "@/components/ui/button";
+import { Dialog, DialogContent, DialogHeader, DialogTitle } from "@/components/ui/dialog";
+import {
+  getAllSupportedChains,
+  SUPPORTED_TOKENS,
+  type SupportedToken,
+} from "@/constants/supportedTokens";
+import type { CreateDonationRequest } from "@/hooks/donation/types";
+import { useCreateDonation } from "@/hooks/donation/useCreateDonation";
+import { useCrossChainBalances } from "@/hooks/donation/useCrossChainBalances";
+import { useDonationTransfer } from "@/hooks/useDonationTransfer";
+import { useProjectStore } from "@/store";
+import { DonationType, PaymentMethod } from "@/types/donations";
+import type { DonationPayment } from "@/utilities/donations/donationExecution";
+import { FiatOnrampModal } from "../FiatOnramp/FiatOnrampModal";
+import { TokenSelector } from "../TokenSelector";
+import { PaymentMethodSelector } from "./PaymentMethodSelector";
+import type { SingleProjectDonateModalProps } from "./types";
+
+function resolvePayoutAddress(
+  payoutAddress: Hex | string | Record<string, string> | undefined,
+  communityContract?: string
+): string {
+  if (!payoutAddress) return "";
+
+  if (typeof payoutAddress === "string") {
+    return isAddress(payoutAddress) ? payoutAddress : "";
+  }
+
+  if (typeof payoutAddress === "object") {
+    if (communityContract && payoutAddress[communityContract]) {
+      const addr = payoutAddress[communityContract];
+      return typeof addr === "string" && isAddress(addr) ? addr : "";
+    }
+
+    const firstValidAddress = Object.values(payoutAddress).find(
+      (addr) => typeof addr === "string" && isAddress(addr)
+    );
+    return firstValidAddress || "";
+  }
+
+  return "";
+}
+
```

Comment:
this function is doing too much, the payoutAddress has a lot of types, better to have only one type